compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -895,9 +895,13 @@ private boolean isAssignable(BType source, BType target, Set<TypePair> unresolve
             return isFiniteTypeAssignable((BFiniteType) source, target, unresolvedTypes);
         }
 
-        if ((targetTag == TypeTags.UNION || sourceTag == TypeTags.UNION) &&
-                isAssignableToUnionType(source, target, unresolvedTypes)) {
-            return true;
+        if ((targetTag == TypeTags.UNION || sourceTag == TypeTags.UNION)) {
+            if (isAssignableToUnionType(source, target, unresolvedTypes)) {
+                return true;
+            }
+
+            TypePair pair = new TypePair(source, target);
+            unresolvedTypes.remove(pair);
```

Comment:
I think that it will be more clear if we can do this on the places where we return false in function `isAssignableToUnionType`

Reply:
Updated. 038381060edf8bc011bfcde35a826687cd830d25